Source/WebKit/UIProcess/Extensions/gtk/WebExtensionContextGtk.cpp

```diff
@@ -0,0 +1,703 @@
+#else
+    ASSERT(webView == m_backgroundWebView);
+#endif
+
+    WebKitNavigationAction* action = webkit_navigation_policy_decision_get_navigation_action(navigationPolicy);
+    RefPtr navigationAction = webkitNavigationActionGetAction(action);
+
+    auto url = URL { String::fromUTF8(webkit_uri_request_get_uri(webkit_navigation_action_get_request(action))) };
+    if (!navigationAction->targetFrame()->isMainFrame() || isURLForThisExtension(url))
+        return true;
+
+    return false;
+}
+
+void WebExtensionContext::didFinishDocumentLoad(WebKitWebView *webView)
+{
+    if (webView != m_backgroundWebView)
+        return;
+
+    // The service worker will notify the load via a completion handler instead.
+    if (protectedExtension()->backgroundContentIsServiceWorker())
+        return;
+
+    performTasksAfterBackgroundContentLoads();
+}
+
+void WebExtensionContext::didFailNavigation(WebKitWebView *webView, RefPtr<API::Error> error)
+{
+    if (webView != m_backgroundWebView)
+        return;
+
+    m_backgroundContentLoadError = createError(Error::BackgroundContentFailedToLoad, nullString(), error);
+    recordErrorIfNeeded(backgroundContentLoadError());
+
+    unloadBackgroundWebView();
+}
+
+void WebExtensionContext::webViewWebContentProcessDidTerminate(WebKitWebView *webView)
+{
+    if (webView == m_backgroundWebView) {
+        unloadBackgroundWebView();
+
+        if (protectedExtension()->backgroundContentIsPersistent())
+            loadBackgroundWebView();
+
+        return;
+    }
+
+#if ENABLE(INSPECTOR_EXTENSIONS)
+    if (isInspectorBackgroundPage(webView)) {
+        webkit_web_view_load_request(m_backgroundWebView.get(), webkit_uri_request_new(inspectorBackgroundPageURL().utf8().data()));
+        return;
+    }
+#endif
+
+    ASSERT_NOT_REACHED();
+}
+
+#if ENABLE(INSPECTOR_EXTENSIONS)
+
+HashSet<Ref<WebProcessProxy>> WebExtensionContext::processes(const API::InspectorExtension& inspectorExtension) const
+{
+    ASSERT(isLoaded());
+    ASSERT(protectedExtension()->hasInspectorBackgroundPage());
+
+    HashSet<Ref<WebProcessProxy>> result;
+
+    RefPtr inspectorProxy = inspector(inspectorExtension);
+    if (!inspectorProxy)
+        return result;
+
+    ASSERT(m_inspectorContextMap.contains(*inspectorProxy));
+
+    const auto& inspectorContext = m_inspectorContextMap.get(*inspectorProxy);
+    if (auto *backgroundWebView = inspectorContext.backgroundWebView.get()) {
+        RefPtr pageProxy = webkitWebViewGetPage(backgroundWebView);
+        result.add(pageProxy->siteIsolatedProcess());
+    }
+
+    return result;
+}
+
+#endif // ENABLE(INSPECTOR_EXTENSIONS)
+
+String WebExtensionContext::declarativeNetRequestContentRuleListFilePath()
+{
+    if (!m_declarativeNetRequestContentRuleListFilePath.isEmpty())
+        return m_declarativeNetRequestContentRuleListFilePath;
+
+    auto directoryPath = storageIsPersistent() ? storageDirectory() : String(FileSystem::createTemporaryDirectory("DeclarativeNetRequest"_s));
```

Comment:
The `createTemporaryDirectory` API you added can silently fail and return an empty string. This seems not desirable.

Reply:
There's another createTemporaryDirectory API that doesn't accept a directory prefix. It can also fail silently and return an empty string. This seems to be standard behaviour for the filesystem functions